### ISSUE TITLE
Disable tracking cookie?

### ISSUE THREAD

**Issue by ziburski** [usefathom/fathom#40]
Hi,

is there a way to disable the tracking cookie completely? I’m not really interested in the returning visitor metric and would prefer to avoid tracking users completely.

Generally, with privacy being a focus for Fathom, it might be nice to have a list (or even a choice) of what exactly is tracked, how it is tracked, and for how long it is stored. 

Cheers!

**Comment by dannyvankooten** [comment]
Hey @ziburski,

That list is definitely on the list - but let me quickly summarise it here right now too as not much is tracked at all:

- No personally identifiable information is tracked. `0`, nada, nothing, nil.
- The only thing that is tracked is a random generated string and information about the last page you visited.
    - Whenever you visit a new page, data about your previous page visit is removed within 1 minute.
    - If you don't visit any new pages, data about your page visit is removed within 30 minutes.

This means that at most, your database only contains information about each visitor's last pageview and that piece of information is removed within 30 minutes (max).

Here's what a single piece of that data may look like:

```
{
  "page": "/contact",
  "session_id": "the-random-string",
  "timestamp": "2018-05-14 10:05:04",
   "is_bounce": 0,
   "is_new_visitor": 1,
   "is_new_session": 1,
}
```

Now the interesting part is that Fathom can only store this little information about your visitors *because* of the "tracking" cookie. Basically we use the visitor's browser as a storage for all visitor-specific information so that we don't have to store it on your server.

I think we may be able to offer the option to _not_ use the tracking cookie but besides slightly simplifying your cookie policy, I'm not too sure on the actual benefits of this to be honest.

**Comment by ziburski** [comment]
Thanks for the thorough explanation, that sounds very reasonable.

Ultimately, I would just love a solution where you don’t need any cookie notices, user consent, or opt-out buttons. The trade-off in UX and effort just isn’t worth it to me, especially for small websites with no existing cookie policy. For users, it’s hard to tell the difference between cookies that protect your privacy and ones that sell you out.

GDPR shouldn’t be a problem for Fathom since it doesn’t track personal data. But the EU cookie law still applies right now, doesn’t it?

**Comment by da2x** [comment]
I'm not a lawyer. This isn't legal advice.

There is no "EU cookie law". The ePrivacy Directive has been interpreted and implemented completely differently by every member state. Some require opt-in before any use of cookies, some require that users are presented with an opt-out, some say that browser settings are enough, ... . You have to comply with the local regulation where your business is based.

The current drat of the ePrivacy Regulation, expected in 2018-Q3, streamlines all of this and say websites must comply with the GDPR (no personal data without opt-in), provide information about cookies, and then align every country with the "browser settings interpretation". Fathom already supports the Do-Not-Track setting and related APIs for site-specific exceptions. Fathom already supports cookie settings as set in the browser. Website publishers just need to document cookie usage and avoid recording any personal data (including username in URLs, etc.).

@dannyvankooten maybe add a FAQ section on usefathom.com?

**Comment by ckluis** [comment]
Adding a "cookie notice" directly into fathom would allow for a single script inclusion instead of wrapping fathom inside another cookie notice inclusion script.    The do-not-track setting doesn't remove the need to request consent from users.

**Comment by da2x** [comment]
Not a lawyer. Not legal advice.

Wouldn’t that lead to duplicate cookie-notice-notices?

There is no single cookie notice that covers all the different local interpretations of the ePrivacy Directive. E.g. in Norway, you don’t even need to display a cookie notice at all! You just need to have all the details in a human-readable way in your privacy policy. For what it’s worth, the current draft for the upcoming ePrivacy Regulation removes ambiguity and uses the Norway-interpretation everywhere. So there **won’t be any reason to include a cookie notice** anymore. Data collection and sharing about individuals (something Fathom doesn’t do) will still require opt-in consent (and not just a notice) to comply with the General Data Protection Regulation.

While we wait for the ePrivacy Regulation, websites have to comply with their local laws which are based on local adaptations of the ePrivacy Directive. Requirements varies greatly from country to country. There is no single notice that works for every European country. So, … the best thing for Fathom to do is to sit and wait on the ePrivacy Regulation.

In summary:
* websites must update their privacy policies, display notices, provide opt-out, or require explicit opt-ins as required by local laws.

Fathom’s job in all of this is to inform publishers what data is collected and how, so that publishers know what to include in their privacy policies and notification mechanisms (when required).

**Comment by ckluis** [comment]
I think you could increase adoption dramatically if you provided the cookie notification tracking directly.  Don't get me wrong - I'll probably use something like https://github.com/snipsco/yett and put the two together, but at the very least providing an integrated solution could be a "Pro" or "Subscription" tier if you want to go down that route.

**Comment by ziburski** [comment]
All the discussion about what constitutes a proper cookie notice seems to validate my suggestion of having the option to remove cookies from Fathom completely.

If Fathom makes tracking simple, I shouldn’t have to read through legal documents from different countries so that I can write a privacy policy that may or may not hold up. I just want to track page views, without making the experience worse for me or my users.

**Comment by dannyvankooten** [comment]
Agree, but we should also not rule out an entire (and valid) technology and actually go against the intent of the law by making privacy worse. 

Without the cookie, Fathom would have to store a whole lot more information about your visitors to get to the data we actually care about. In this case, a cookie (and the same applies to other client-side storage mechanisms) is the most privacy friendly option out there and probably therefore the easiest method to comply with the law.

I'm sorry, but I don't see how we could get rid of the cookie/client-side storage without making privacy worse. That said, we are drafting up an article explaining what the cookie is for and why it is, by far, the easiest option. 

I'm also not a lawyer and this is not legal advice, but personally I would feel very confident in not asking for explicit consent for the Fathom cookie, given the current laws. I would still document its use and benefits, and we're working on making that easier for you.

**Comment by driv3r** [comment]
just an idea - what if it would be possible to make the cookie storage as an adapter? I was thinking, and have a working draft, that you could just store the information in the global variable, which will then hold the info during single real "page view". Today when most of the websites are SPAs, or static pages using tools like turbolinks, the page is never reloaded, thus you can see if user is browsing through the website, within single "page view".

I think it's a good compromise between privacy notices, counts of returning visits and keeping tracking js simple. WDYS?

**Comment by laander** [comment]
@driv3r Good suggestion, but without any type of session tracking between reloads (client or server side), a browser reload would track as a new visitor.

In some cases, for SPAs, that _might_ be acceptable, but for classic websites you'd end up with an extremely inflated visitors count. The option could be provided as an optional flag though, but I still don't think the trade-off is worth it.

**Comment by jlelse** [comment]
Currently Fathom shows information about page views, unique visitors and referrers. Without cookies page views and referrers should be possible and losing information about unique visitors is probably worth out when you just want to see which pages are most visited.

**Comment by jhabdas** [comment]
Cookies are sent with all requests to the origin, so images etc will be collected once the cookie is placed – that could be a problem if a proxy is used to secretly send data on the same origin to the likes of DoubleClick. I don't believe there should be an adapter here – there simply shouldn't be a cookie. 

Instead I'd suggest fathom use `sessionStorage` and remove the cookie at priority.

**Comment by jhabdas** [comment]
@dannyvankooten I saw the `wontfix` tag added to this issue. I read your explainer and feel you're trying to justify an architectural mistake. No offense, but ff you don't intend to resolve this issue I'm confident a fork will.

**Comment by jhabdas** [comment]
Cookies are sent with all requests to the origin, so images etc will be collected once the cookie is placed – that could be a problem if a proxy is used to secretly send data on the same origin to the likes of DoubleClick. I don't believe there should be an adapter here – there simply shouldn't be a cookie. 

Instead I'd suggest fathom use `sessionStorage` and remove the cookie at priority.

@dannyvankooten I saw the `wontfix` tag added to this issue. I read your explainer and feel you're trying to justify an architectural mistake. No offense, but if you don't intend to resolve this issue I'm confident someone else will.

**Comment by JackEllis** [comment]
@jhabdas I don't think sessionStorage is a realistic solution here, as it empties as soon as you close the tab. This doesn't help with tracking unique visitors since each new tab would be a fresh sessionStorage. Personally, I'd rather not send the cookies on each request and I understand where you're coming from. I think that something like localStorage may end up being useful for this in the future.

**Comment by jhabdas** [comment]
> I think that something like `localStorage` may end up being useful for this in the future.

I'd settle for `localStorage`. There's just too much CloudFlare et al. can do with cookies these days despite our best attempts or good intentions. For instance, I just saw today the old Ledger Manager extension was doing a sidecar cookie off-domain for anyone visiting DevDocs.io ([proof](https://github.com/freeCodeCamp/devdocs/issues/971)). Ugh.

If repeat visitor data is important to Fathom it would be nice to offer an cookie-free option for those of us who really want a privacy-focused analytics solution until HTTP3 starts tracking us at the protocol layer.

**Comment by EricKit** [comment]
Another reason for `localStorage` is for those using platforms such as Ionic on an iPhone where Safari runs in the app. Cookies are not stored so each page view is a new user and a new session.

**Comment by EricKit** [comment]
This is disappointing for all mobile framework users that can't use cookies.

**Comment by JackEllis** [comment]
Sorry, I forgot to add a comment, it’s on our list!

**Comment by EricKit** [comment]
What I meant to say based on this new information is that "this is very exciting for all mobile framework users that can't use cookies." Thanks!

**Comment by Spone** [comment]
@JackEllis any news on this?

**Comment by kaishin** [comment]
I personally don't care about unique and returning visitors. Please reconsider this for the Lite version. I can't recommend Fathom to my clients, even if they go Pro, if I don't see some good faith in fixing this for an analytics service that's focused on privacy. @dannyvankooten

**Comment by JackEllis** [comment]
We don’t use tracking cookies on pro! Lite will follow I expect!

**Comment by kaishin** [comment]
> We don’t use tracking cookies on pro! Lite will follow I expect!
That I am aware of. Looking forward to this change coming to Lite as well.

 I deleted the original comment after I realized that it might come off the wrong way. But since you replied I would like to reiterate that this is very important for me.

**Comment by JackEllis** [comment]
@kaishin Noted, thanks Reda
